src/core/tsi/ssl_transport_security.cc

```diff
@@ -1450,9 +1523,22 @@ static tsi_result ssl_handshaker_next(
   if (ssl_handshaker_get_result(impl) == TSI_HANDSHAKE_IN_PROGRESS) {
     *handshaker_result = nullptr;
   } else {
-    size_t unused_bytes_size = received_bytes_size - bytes_consumed;
-    const unsigned char* unused_bytes =
-        unused_bytes_size == 0 ? nullptr : received_bytes + bytes_consumed;
+    // In TLS 1.3, the client may send application data records in the same
+    // flight of messages as the record containing the ClientFinished message.
+    // In TLS 1.2, this is not allowed; both the client and server must complete
+    // the handshake before any application data may be sent.
```

Comment:
This still looks off. On a TLS 1.2 resumption, the client speaks last, so you can get application data and client Finished in the same flight. TLS 1.2 full handshakes do end with a server flight, but all browsers implement an optimization where they [send data earlier anyway](https://tools.ietf.org/html/rfc7918) for some connections, notably all HTTP/2 ones.

And, conversely, in any flow where the server Finished is last, the server may send application data immediately after its handshake message.

Reply:
Thanks for the explanation David! In a reply to your other comment, I explain what the gRPC issue was that I was trying to address. Once we settle that, I'll come back and revise the comment.